core: Defer on transitive dependencies for data resources with conditions (#31028)
hashicorp/terraform: website/docs/language/expressions/custom-conditions.mdx

- Terraform evaluates `precondition` blocks after evaluating existing `count` and `for_each` arguments. This lets Terraform evaluate the precondition separately for each instance and then make `each.key`, `count.index`, etc. available to those conditions. Terraform also evaluates preconditions before evaluating the resource's configuration arguments. Preconditions can take precedence over argument evaluation errors.
- Terraform evaluates `postcondition` blocks after planning and applying changes to a managed resource, or after reading from a data source. Postcondition failures prevent changes to other resources that depend on the failing resource.

> In most cases we do not recommend including both a `data` block and a `resource` block that both represent the same object in the same configuration, because that can prevent Terraform from understanding that the `data` block result can be affected by changes in the `resource` block. However, in situations where you need to check a result of a `resource` block that the resource itself doesn't directly export, you can use a `data` block to check that object safely as long as you place the check as a direct `postcondition` of the `data` block. Associating a postcondition with a data resource tells Terraform that the `data` block is serving as a check of an object defined elsewhere, allowing Terraform to perform actions in the correct order.

Reviewer: A couple of nits to try to make this the tiniest bit shorter/easier to read. Also, I wanted to remove the `` around postcondition because there are so many other things that have code formatting in this paragraph and sometimes that can make things harder to read. I think it's okay here because we don't say "postcondition block" and are referring to postcondition as a concept. Push back if you hate this, but that's why I tried it this way :) 

```suggestion
In most cases, we do not recommend including both a `data` block and a `resource` block that represent the same object in the same configuration. This can prevent Terraform from understanding that the `data` block result can be affected by changes in the `resource` block. However, when you need to check a result of a `resource` block that the resource itself does not directly export, you can use a `data` block to check that object safely as long as you place the check as a direct postcondition of the `data` block. This tells Terraform that the `data` block is serving as a check of an object defined elsewhere, allowing Terraform to perform actions in the correct order.
```

    name = "volume-id"
    values = [aws_instance.example.root_block_device.volume_id]
  }

>   # Whenever a data resource is verifying the result of a managed resource
>   # declared in the same configuration, you MUST write the checks as
>   # postconditions of the data resource so that Terraform can see that
>   # the data resource read must wait until after any changes to the
>   # managed resource have completed.

Reviewer: ```suggestion
  # Whenever a data resource is verifying the result of a managed resource
  # declared in the same configuration, you MUST write the checks as
  # postconditions of the data resource. This ensures Terraform will wait
  # to read the data resource until after any changes to the
  # managed resource have completed.
```
